samples/grid/griddemo.cpp

```diff
@@ -1235,31 +1235,91 @@ void GridFrame::SetGridLineColour( wxCommandEvent& WXUNUSED(ev) )
     }
 }
 
+namespace
+{
+
+// Used when inserting/deleting rows/columns to allow changing by more than
+// one row/column at once. Note that the passed array is already sorted.
+bool IsConsecutiveSelection(const wxArrayInt& selection)
+{
+    const int selCount = static_cast<int>( selection.size() );
+
+    if ( !selCount )
+        return false;
+
+    const int first = selection[0];
+
+    for ( int i = 1; i < selCount; ++i )
+    {
+        if ( selection[i] != first + i )
+            return false;
+    }
+
+    return true;
+}
+
+} // anoymous namespace
 
 void GridFrame::InsertRow( wxCommandEvent& WXUNUSED(ev) )
 {
-    grid->InsertRows( grid->GetGridCursorRow(), 1 );
+    int pos, insertions;
+
+    const wxArrayInt selected = grid->GetSelectedRows();
```

Comment:
I was going to advise the following micro-optimization:
```suggestion
    const wxArrayInt& selected = grid->GetSelectedRows();
```
to avoid making an extra copy of a potentially big array, but then I realized that this actually shows a wrong example from efficiency point of view: if you're only interested in whether only consecutive rows are selected, you should use `GetSelectedBlocks()` and check that it has a single element, i.e.
```cpp
        const wxGridBlocks selected = m_grid->GetSelectedBlocks();
        wxGridBlocks::iterator it = selected.begin();
        if ( ++it == selected.end() )
            ... consecutive ...
```
(maybe we should also add `wxGridBlocks::GetCount()` to simplify this...)

This allows to avoid splitting the selected block(s) into rows just to check if they recombine into a single block here.

The same comment applies to all the other places where `IsConsecutiveSelection()` is used too, of course.

Reply:
> 
> ```c++
>         const wxGridBlocks selected = m_grid->GetSelectedBlocks();
>         wxGridBlocks::iterator it = selected.begin();
>         if ( ++it == selected.end() )
>             ... consecutive ...
> ```

Interesting suggestion, thanks!


> (maybe we should also add `wxGridBlocks::GetCount()` to simplify this...)

Not sure if `GetCount()` would be useful here at least, usually you want to do something with the iterator such as in this case. I was only interested in consecutive selections to do multiple insertions or deletions at once instead of only by one per insert/delete call.

> The same comment applies to all the other places where IsConsecutiveSelection() is used too, of course.

This would make the difference between how inserting and deleting works in the sample bigger. Currently row insertions work from either the grid cursor or you need a row selection (not cell selection) for multiple insertions. Deletion always requires a row selection.

I think what deletion requires is better (and more like how a spreadsheet program seems to work), but needing just a selection as you propose would be slightly more convenient for testing.

However if deletion should work based on block selections: how to then delete the selected rows (from all blocks) without resolving to what I think would be quite some extra code? If requiring row selection maybe `GetSelectedRowBlocks` could help.

Note that the current sample (also master) for deletion only works on row or column selections but it may appear to support based on cell selection: when I select cell B1 (not merely activate) then I can't delete selected rows, however deleting selected columns does work!?

When I move to B2 (in essence: away from the top row) then deleting the column also doesn't work. 

Conversely the same is true with columns: Selecting cell A2: can't delete column, can delete row. Move a cell selection away from the left-most column and deleting won't work again.

Not sure if that's by design and just unexpected for me, or a problem with the sample or library.

Reply:
Sorry, I didn't want to open a can of worms here... I thought `IsConsecutiveSelection()` could be just trivially replaced with the test `GetSelectedBlocks().GetCount() == 1`, but I didn't really check it. I am still not sure in which situation does the return value of `IsConsecutiveSelection()` differ from this expression however?

The behaviour in the sample is surprising, I don't know why does it happen neither, but I think it's something that the sample rather than the library itself does, but, again, I didn't have time to actually check it, so I could be wrong.

Reply:
> Sorry, I didn't want to open a can of worms here...

It was still closed when passed at least. And sorry for opening another one (which hopefully can either be discarded or go to a ticket).

> I thought `IsConsecutiveSelection()` could be just trivially replaced with the test `GetSelectedBlocks().GetCount() == 1`, but I didn't really check it. I am still not sure in which situation does the return value of `IsConsecutiveSelection()` differ from this expression however?

`IsConsecutiveSelection()` gets passed an array from `GetSelectedRows()` which works only on row selections, and there's no row selection when I merely select cells. While the expression with `GetSelectedBlocks()` works with cell selections.
So now insertions can work with just a cell selection (as mentioned, convenient but maybe not expected) but the code for deletion still requires row selections. That seems a strange difference to me and I think insertion and deletion should both either work based on row/column selections or cell selections. I'll commit and push the wip changes.

Trying to then make deletion work based on `GetSelectedBlocks()` looked problematic as mentioned. I now also checked `GetSelectedRowBlocks()` but that only returns something in `wxGridSelectRows` mode. This seems to be by design, 3307000(baad!): "[...] using wxGrid in row- or column-only selection mode". After seeing `enum wxGridSelectionModes` with bit flags (a somewhat curious combination) I wonder if this behaviour is desired. After all the grid sample starts in `wxGridSelectCells` mode which "allow[s] selecting anything", including then entire rows. 'Even' with `wxGridSelectRowsOrColumns` mode, `GetSelectedRowBlocks()` also returns an empty vector. Brute-forcing the latter to work in other modes seems to do more of what I would expect: It returns blocks when there's a row selection, just like `GetSelectedRows()` returns something. However it also returns rows based on cell selections which it probably shouldn't, similar to `GetSelectedRows()`.

If that's something to actually change (regardless if it's needed for the sample) I could open a ticket. Some interesting helper function would probably be needed for checking `wxGridSelectionModes` flags. And if `GetSelectedRowBlocks()` should only return something if selecting entire rows (and not just cells, as it could easily do now) then I suppose more code is needed beyond my current knowledge of limited exploring.


> The behaviour in the sample is surprising, I don't know why does it happen neither, but I think it's something that the sample rather than the library itself does, but, again, I didn't have time to actually check it, so I could be wrong.

I'm afraid it's the library, will try to figure out at least where the difference comes from.

Reply:
> > The behaviour in the sample is surprising, I don't know why does it happen neither, but I think it's something that the sample rather than the library itself does, but, again, I didn't have time to actually check it, so I could be wrong.
> 
> I'm afraid it's the library, will try to figure out at least where the difference comes from.

It's clearly the sample as it uses `IsInSelection(n, 0)` to determine deletion (hence a single cell selection in column 0 qualifies, but not in another column) instead of making use of `GetSelectedRows()` in master. Sorry for the false alarm.

Reply:
> `IsConsecutiveSelection()` gets passed an array from `GetSelectedRows()` which works only on row selections, and there's no row selection when I merely select cells. While the expression with `GetSelectedBlocks()` works with cell selections.

I see, thanks. In principle, I'd say that this just proves that `GetSelectedBlocks()` is the right function to use because it clearly shouldn't matter how the entire row (or several consecutive rows) got selected, what matters is that they're selected. I.e. I think the sample just does something weird now.

>  I now also checked `GetSelectedRowBlocks()` but that only returns something in `wxGridSelectRows` mode. This seems to be by design

Yes, it was. I thought they wouldn't be useful in other modes, i.e. what would be the advantage compared to `GetSelectedBlocks()` itself? And I admit I still don't see it...

> 'Even' with `wxGridSelectRowsOrColumns` mode, `GetSelectedRowBlocks()` also returns an empty vector.

... except in this one where it probably can be useful, so it probably should work in it too. OTOH I find `wxGridSelectRowsOrColumns` mode pretty useless on its own, so let's just say it's not the highest priority.

Reply:
> > `IsConsecutiveSelection()` gets passed an array from `GetSelectedRows()` which works only on row selections, and there's no row selection when I merely select cells. While the expression with `GetSelectedBlocks()` works with cell selections.
> 
> I see, thanks. In principle, I'd say that this just proves that `GetSelectedBlocks()` is the right function to use because it clearly shouldn't matter how the entire row (or several consecutive rows) got selected, what matters is that they're selected. I.e. I think the sample just does something weird now.

This could be terminology. Talking about the `wxGridSelectCells` mode: For me a row is selected if I click on its header. If I select a single cell then its row is not selected, I suppose it is partially but not entirely. This corresponds with `GetSelectedRows()`: there is no row selection if I select only cells.

> Yes, it was. I thought they wouldn't be useful in other modes,

Maybe I'm missing what both are trying to do, but then why is it useful for `GetSelectedRows()` to return something in `wxGridSelectCells` mode? I don't know what the function does for each mode, but I expected `GetSelectedRowBlocks()` to behave the same (perhaps because of the similarity in name).

> i.e. what would be the advantage compared to `GetSelectedBlocks()` itself? And I admit I still don't see it...

What if I want to get the selected rows (as in my definition from above, so no rows are selected when selecting cells) ? Am I supposed to filter them from `GetSelectedBlocks()` by checking the column span? And additionally for (reverse) deletion, sort the rows?
`GetSelectedRowBlocks()` seemed like it could help with both, if it behaved like `GetSelectedRows()`.

Reply:
> This could be terminology. Talking about the `wxGridSelectCells` mode: For me a row is selected if I click on its header. If I select a single cell then its row is not selected, I suppose it is partially but not entirely.

Yes, of course, if you select a single cell then the row is not selected. The question is what happens if you select _all_ the cells in a row one by one. IMO it's more useful to consider the row as being selected, even if it hasn't been selected all at once.

> This corresponds with `GetSelectedRows()`: there is no row selection if I select only cells.

IMO it's a problem or even a bug.

> Maybe I'm missing what both are trying to do, but then why is it useful for `GetSelectedRows()` to return something in `wxGridSelectCells` mode? I don't know what the function does for each mode, but I expected `GetSelectedRowBlocks()` to behave the same (perhaps because of the similarity in name).

As usual, there is also the question of backwards compatibility. `GetSelectedRows()` is an old function and we can't (and hopefully didn't) change its behaviour. `GetSelectedRowBlocks()` is the new function which was mostly added for efficiency reasons: `GetSelectedRows()` is unusable for large grids, while `GetSelectedRowBlocks()` definitely is.

> What if I want to get the selected rows (as in my definition from above, so no rows are selected when selecting cells) ?

Sorry, I just don't think it's the right thing to do. If you don't want to allow selecting the individual cells at all, just use the row selection mode. If you do want to allow selecting the individual cells, then it doesn't make sense to distinguish between selecting the row all at once or by selecting its individual cells because they both appear the same to the user -- so why should they behave differently?

> Am I supposed to filter them from `GetSelectedBlocks()` by checking the column span? And additionally for (reverse) deletion, sort the rows?

No, of course not.

> `GetSelectedRowBlocks()` seemed like it could help with both, if it behaved like `GetSelectedRows()`.

I think that `GetSelectedRowBlocks()` behaviour is much more useful than that of `GetSelectedRows()`.

Reply:
> Yes, of course, if you select a single cell then the row is not selected. The question is what happens if you select _all_ the cells in a row one by one. IMO it's more useful to consider the row as being selected, even if it hasn't been selected all at once.

Definitely, it shouldn't matter how they all got selected just that they are. And it's something `GetSelectedBlocks()` can help with, while with `GetSelectedRows()` the row is only selected if there's a single selection spanning the entire row (or clicking on its header, but that's pretty much the same under the hood).

> > This corresponds with `GetSelectedRows()`: there is no row selection if I select only cells.
> 
> IMO it's a problem or even a bug.

Reading this and some later statements I see where this went wrong now and I should have been more clear: with "cells" I meant (one,) a few or many in a grid. The only "cells" I didn't mean were all of them in a row to form an entire row span. In that latter case there had better be a row selection. Sorry for the confusion, and at least we were agreeing here...

> As usual, there is also the question of backwards compatibility. `GetSelectedRows()` is an old function and we can't (and hopefully didn't) change its behaviour. `GetSelectedRowBlocks()` is the new function which was mostly added for efficiency reasons: `GetSelectedRows()` is unusable for large grids, while `GetSelectedRowBlocks()` definitely is.

I don't know why `GetSelectedRowBlocks()` wouldn't continue that convenient behaviour of returning rows in `wxGridSelectCells` mode, it would help with it being a replacement for users wanting to use something more efficient than `GetSelectedRows()`. Regardless of whether it's useful, wxGrid has been promoting row and column selections in `wxGridSelectCells` both [in](https://github.com/wxWidgets/wxWidgets/blob/414f211b8a21ade4d7850e163070386647b4e9e0/src/generic/grid.cpp#L3741-L3745) [code](https://github.com/wxWidgets/wxWidgets/blob/414f211b8a21ade4d7850e163070386647b4e9e0/src/generic/gridsel.cpp#L169-L173) and [docs](https://github.com/wxWidgets/wxWidgets/blob/414f211b8a21ade4d7850e163070386647b4e9e0/interface/wx/grid.h#L2914-L2917).
I don't find it strange that I can do `SelectRow()` in cell mode (it probably would have been if there was something like a cells-only mode), but that I then get a different row selection depending on using `GetSelectedRows()` or `GetSelectedRowBlocks()` is unexpected to me. I don't see having new functions as a good enough reason to change behaviour here, rather it seems confusing. If I don't expect rows to be selected in cell mode then I shouldn't make use of row selection functions that could return them, so I don't think there would even be a use for an additional `GetSelectedRowBlocksDespiteBeingInSelectCellsMode()` convenience function and instead the current new ones perhaps should change.

> > Am I supposed to filter them from `GetSelectedBlocks()` by checking the column span? And additionally for (reverse) deletion, sort the rows?
> 
> No, of course not.

For basing deletions on only selected rows I think it would have been. However both insertion and deletion are just based on `GetSelectedBlocks()` now. With my previous changes it was a confusing mix of behaviour based on either `GetSelectedRows()` for the new consecutive deletions or otherwise `IsInSelection()`.

Reply:
I had initially decided to only handle `wxGridSelectRows` mode in `GetSelectedRowBlocks()` because there was no ambiguity about what it should do in this mode and I wasn't totally sure about the other ones and, to be frank, didn't think it was worth spending time thinking about it. However now that you've already done all the thinking, I think I agree and wouldn't mind if we change it as you propose. But this is a pure enhancement, i.e. it can be done later, without affecting any existing code, so it's not exactly the highest priority neither.

IOW let's merge this one, which fixes a real problem for now, but we can improve `GetSelected{Row,Col}Blocks()` later, of course.

Reply:
Sure, thanks and just an enhancement indeed. I don't see how it could get in the way and it's probably expected by long-time grid users making use of 'mixed' selections mode in some form. I could look into adding more repeated lines for the allowed modes or refactoring much later.